File: server/src/services/bingoLogic.ts
import type { BingoCard, CardCell, LinePattern } from "../types.js";

const COLUMN_RANGES: Record<string, number[]> = {
  B: Array.from({ length: 15 }, (_, i) => i + 1),
  I: Array.from({ length: 15 }, (_, i) => i + 16),
  N: Array.from({ length: 15 }, (_, i) => i + 31),
  G: Array.from({ length: 15 }, (_, i) => i + 46),
  O: Array.from({ length: 15 }, (_, i) => i + 61),
};

function sample<T>(items: T[], count: number): T[] {
  const copy = [...items];
  const picks: T[] = [];
  for (let i = 0; i < count; i += 1) {
    const index = Math.floor(Math.random() * copy.length);
    picks.push(copy[index]!);
    copy.splice(index, 1);
  }
  return picks;
}

export function generateBingoCard(): BingoCard {
  const columns: CardCell[][] = [];
  for (const [letter, numberRange] of Object.entries(COLUMN_RANGES)) {
    const picks = sample(numberRange, 5).sort((a, b) => a - b);
    columns.push(
      picks.map((value) => ({
        letter,
        value,
        is_free: false,
        marked: false,
      })),
    );
  }

  columns[2]![2] = { letter: "N", value: null, is_free: true, marked: true };
  const grid = Array.from({ length: 5 }, (_, rowIdx) =>
    Array.from({ length: 5 }, (_, colIdx) => columns[colIdx]![rowIdx]!),
  );
  return { grid };
}

export function generateDrawSequence(): number[] {
  const numbers = Array.from({ length: 75 }, (_, i) => i + 1);
  for (let i = numbers.length - 1; i > 0; i -= 1) {
    const j = Math.floor(Math.random() * (i + 1));
    [numbers[i], numbers[j]] = [numbers[j]!, numbers[i]!];
  }
  return numbers;
}

export function markCard(card: BingoCard, drawnNumber: number): BingoCard {
  for (const row of card.grid) {
    for (const cell of row) {
      if (cell.value === drawnNumber) {
        cell.marked = true;
      }
    }
  }
  return card;
}

export function checkWin(
  card: BingoCard,
  winningLineTarget = 1,
  allowedLinePatterns: LinePattern[] | null = null,
  allowFullHouse = true,
): { winner: boolean; pattern: string | null } {
  const grid = card.grid;
  const lineTarget = Math.max(1, winningLineTarget);
  const allowed = new Set(allowedLinePatterns ?? ["horizontal", "vertical", "diagonal"]);

  if (allowFullHouse && grid.every((row) => row.every((cell) => cell.marked))) {
    return { winner: true, pattern: "full_house" };
  }

  let horizontalLines = 0;
  let verticalLines = 0;
  let diagonalLines = 0;

  if (allowed.has("horizontal")) {
    for (const row of grid) {
      if (row.every((cell) => cell.marked)) horizontalLines += 1;
    }
  }

  if (allowed.has("vertical")) {
    for (let col = 0; col < 5; col += 1) {
      if (grid.every((row) => row[col]!.marked)) verticalLines += 1;
    }
  }

  if (allowed.has("diagonal")) {
    if (grid.every((row, idx) => row[idx]!.marked)) diagonalLines += 1;
    if (grid.every((row, idx) => row[4 - idx]!.marked)) diagonalLines += 1;
  }

  const totalLines = horizontalLines + verticalLines + diagonalLines;
  if (totalLines < lineTarget) {
    return { winner: false, pattern: null };
  }
  if (horizontalLines >= lineTarget) return { winner: true, pattern: "horizontal" };
  if (verticalLines >= lineTarget) return { winner: true, pattern: "vertical" };
  if (diagonalLines >= lineTarget) return { winner: true, pattern: "diagonal" };
  return { winner: true, pattern: `${lineTarget}_line_combo` };
}

File: app/services/bingo_logic.py
from __future__ import annotations

import random
import os, sys

cur_dir = os.getcwd()
parent_dir = os.path.realpath(os.path.join(os.path.dirname(cur_dir)))
if parent_dir not in sys.path:
    sys.path.append(parent_dir)
    sys.path.append(cur_dir)
sys.path.insert(1, ".")

from app.models.schemas import BingoCard, CardCell, WinCheckResponse


COLUMN_RANGES = {
    "B": range(1, 16),
    "I": range(16, 31),
    "N": range(31, 46),
    "G": range(46, 61),
    "O": range(61, 76),
}


def generate_bingo_card() -> BingoCard:
    columns: list[list[CardCell]] = []
    for letter, number_range in COLUMN_RANGES.items():
        picks = random.sample(list(number_range), 5)
        column = [
            CardCell(letter=letter, value=value, is_free=False, marked=False)
            for value in sorted(picks)
        ]
        columns.append(column)

    columns[2][2] = CardCell(letter="N", value=None, is_free=True, marked=True)
    grid = [[columns[col_idx][row_idx] for col_idx in range(5)] for row_idx in range(5)]
    return BingoCard(grid=grid)


def generate_draw_sequence() -> list[int]:
    numbers = list(range(1, 76))
    random.shuffle(numbers)
    return numbers


def mark_card(card: BingoCard, drawn_number: int) -> BingoCard:
    for row in card.grid:
        for cell in row:
            if cell.value == drawn_number:
                cell.marked = True
    return card


def check_win(
    card: BingoCard,
    winning_line_target: int = 1,
    allowed_line_patterns: list[str] | None = None,
    allow_full_house: bool = True,
) -> WinCheckResponse:
    grid = card.grid
    line_target = max(1, int(winning_line_target))
    allowed = set(allowed_line_patterns or ["horizontal", "vertical", "diagonal"])

    if allow_full_house and all(cell.marked for row in grid for cell in row):
        return WinCheckResponse(winner=True, pattern="full_house")

    horizontal_lines = 0
    vertical_lines = 0
    diagonal_lines = 0

    if "horizontal" in allowed:
        for row in grid:
            if all(cell.marked for cell in row):
                horizontal_lines += 1

    if "vertical" in allowed:
        for column_index in range(5):
            if all(grid[row_index][column_index].marked for row_index in range(5)):
                vertical_lines += 1

    if "diagonal" in allowed:
        if all(grid[index][index].marked for index in range(5)):
            diagonal_lines += 1

        if all(grid[index][4 - index].marked for index in range(5)):
            diagonal_lines += 1

    total_lines = horizontal_lines + vertical_lines + diagonal_lines
    if total_lines < line_target:
        return WinCheckResponse(winner=False, pattern=None)

    if horizontal_lines >= line_target:
        return WinCheckResponse(winner=True, pattern="horizontal")
    if vertical_lines >= line_target:
        return WinCheckResponse(winner=True, pattern="vertical")
    if diagonal_lines >= line_target:
        return WinCheckResponse(winner=True, pattern="diagonal")

    return WinCheckResponse(winner=True, pattern=f"{line_target}_line_combo")
